### ISSUE TITLE
Lazy dataprovider loading

### ISSUE THREAD

**Issue by SamMousa** [sebastianbergmann/phpunit#3191]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 7.1.5
| PHP version         | 7.2
| Installation Method | Composer

This is not a bug, but more of a feature request / proposal.

Currently the result of a dataprovider is always traversed, meaning that a generator will be fully traversed  before the first test starts.

This approach has several downsides:
- The whole data set must fit in memory.
- Actual testing is delayed.

Changes required would be relatively minor.
- `PHPUnit\Util\Test::getProvidedData()` return type should be changed to `?Traversable`.
- Instead of testing the dataset beforehand we should test each element separately, we can easily do this by wrapping the `Traversable`.
`TestSuite::createTest()` needs some minor changes (can't check if a `Traversable` is empty).

I'm open for implementing if there's agreement that this would be an improvement.

**Comment by sebastianbergmann** [comment]
Depends on #10.

**Comment by keradus** [comment]
note: currently there is a possibility to filter tests by data provider label, this shall not be broken

**Comment by epdenouden** [comment]
> note: currently there is a possibility to filter tests by data provider label, this shall not be broken

Just to clarify, you are referring to `--filter someName` also acting on `@dataprovider` row names or is there some other filtering mechanism I am not yet aware of?

**Comment by keradus** [comment]
> Just to clarify, you are referring to --filter someName also acting on @dataprovider row names or is there some other filtering mechanism I am not yet aware of?

example of data provider:
```php
public function provideFoo() {
    return [
        'foo_bar' => [1, 2],
        'baz_123' => [3, 4],
    ];
}
```

then, you can `--filter foo_bar`

**Comment by epdenouden** [comment]
@keradus Yes, this scenario works in the prototype. It did require some new logic to check if a dataprovider requires being loaded for filtering to work, but yes, this works fine!

**Comment by epdenouden** [comment]
Hooray! I've got some first results as of yesterday:
- saves between 5%-10% even on our own unit self-test suite
- smoother resource usage; much smaller spikes during initialization
- no changes to tests required, everything works as-is
- works with selectors like `--testsuite` and `--filter`

In the graphs below you can see the first results of testing:
- **Prefetched** is the default `@dataprovider` implementation from v8.x
- **JIT** is the MVP just-in-time loading mechanism, including `--filter` support
- **JIT+unload** is the improved MVP that unloads data after testing

`Generator`-based row-by-row loading of data is a logical next step. This will reduce or even eliminate initial load spikes and further smooth out system load.

![image](https://user-images.githubusercontent.com/26651359/59841172-16c42500-9354-11e9-8d76-af6cbb702473.png)

![image](https://user-images.githubusercontent.com/26651359/59841642-0b252e00-9355-11e9-9686-d43503151d93.png)

Notes on testing:
- PHPUnit's unit self-tests only use simple and small dataproviders: no large sets, no complex datatypes, no database interaction
- memory usage during a run of `./phpunit --testsuite unit` with `xdebug.trace_format=1`
- all measurements done on my own Macbook Air, so I've avoided using absolute numbers where possible

**Comment by epdenouden** [comment]
@sebastianbergmann @keradus The MVP works and can be refined to release quality. I'd love to be able to spend some time on this and writing some proper docs + tutorials, if anybody knows of a company looking to sponsor some OSS work. :)

**Comment by keradus** [comment]
> It did require some new logic to check if a dataprovider requires being loaded for filtering to work

out of curiosity, can you elaborate that logic? like, how to detect if dataprovider shall be loaded or not to filter by keys provided by dataprovider ?

**Comment by epdenouden** [comment]
> out of curiosity, can you elaborate that logic? like, how to detect if dataprovider shall be loaded or not to filter by keys provided by dataprovider ?

Unfortunately there is no magic here, I wish. Filtering by name requires the names to be available, which in turn forces us to load the dataprovider. The way this works is by loading the data upon `DataProviderTestSuite::injectFilter` so the data is available for the iterator filter.

There are many optimizations that came to mind, all of them introducing their own edge cases. Any ideas welcome as always! I'm intend on building a proper implementation now, so your critical eye is very much appreciated.

**Comment by epdenouden** [comment]
_if @keradus orders a meal in a restaurant, does the kitchen iterate through the whole menu only to return his order?_

**Comment by epdenouden** [comment]
ping @SamMousa on-demand data providers are a possible :)

**Comment by keradus** [comment]
*if @epdenouden ordered anything with salami, does the kitchen knows where to look for it, or iterate over all dishes ingredients?*

_problematic_ part is that we can filter not only by items (test classes, methods) but also ingredients (case description)

**Comment by epdenouden** [comment]
> _if @epdenouden ordered anything with salami, does the kitchen knows where to look for it, or iterate over all dishes ingredients?_
> 
> _problematic_ part is that we can filter not only by items (test classes, methods) but also ingredients (case description)

It will iterate through al dishes looking for salami as an ingredient. This works by simply reverting to the old data provider behaviour when a `--filter` is used. There are still some efficiency gains possible, but the main advantage of not loading the providers is gone

 It might be possible, though! I am not (yet) familiar enough with the whole test filtering logic.

**Comment by epdenouden** [comment]
Superseded by https://github.com/sebastianbergmann/phpunit/issues/3736

**Comment by keradus** [comment]
> This works by simply reverting to the old data provider behaviour when a --filter is used. There are still some efficiency gains possible, but the main advantage of not loading the providers is gone

It may happen that a lot of great effort made to improve data providers will be simply ignored due to fact that someone used a filter to run only some or the tests ? What a pity, I almost always use filtering, @epdenouden
